pages/api/levels/query.ts

```diff
@@ -0,0 +1,119 @@
+import withAuth, { NextApiRequestWithAuth } from '../../../lib/withAuth';
+
+import Level from '../../../models/db/level';
+import { LevelModel } from '../../../models/mongoose';
+import type { NextApiResponse } from 'next';
+import dbConnect from '../../../lib/dbConnect';
+
+export async function doQuery(query:any) {
+  await dbConnect();
+
+  const { name, username, author_note, min_moves, max_moves, time_range, page, sort_by, sort_dir } = query as {name:string, username: string, author_note:string, min_moves:string, max_moves:string, time_range:string, min_rating:string, page:string, sort_by:string, sort_dir:string};
+  const searchObj = { 'isDraft': false } as {[key:string]:any};
+  const limit = 10;
+
+  if (name) {
+    searchObj['name'] = {
+      $regex: name,
+      $options: 'i',
+    };
+  }
+
+  if (author_note) {
+    searchObj['authorNote'] = {
+      $regex: author_note,
+      $options: 'i',
+    };
+  }
+
+  if (min_moves && max_moves) {
+    searchObj['leastMoves'] = {
+      $gte: parseInt(min_moves),
+      $lt: parseInt(max_moves),
+    };
+  }
+
+  if (time_range) {
+    searchObj['ts'] = {}; // all time
+
+    if (time_range === '24h') {
+      searchObj['ts']['$gte'] = new Date(Date.now() - 24 * 60 * 60 * 1000);
+    }
+    else if (time_range === '7d') {
+      searchObj['ts']['$gte'] = new Date(Date.now() - 7 * 24 * 60 * 60 * 1000);
+    }
+    else if (time_range === '30d') {
+      searchObj['ts']['$gte'] = new Date(Date.now() - 30 * 24 * 60 * 60 * 1000);
+    }
+  }
+
+  const sortObj = {} as {[key:string]:any};
+  const sort_direction = (sort_dir === 'asc') ? 1 : -1;
+
+  if (sort_by) {
+    if (sort_by === 'moves') {
+      sortObj['leastMoves'] = sort_direction;
+    }
+    else if (sort_by === 'ts') {
+      sortObj['ts'] = sort_direction;
+    }
+    else if (sort_by === 'reviews_score') {
+      sortObj['calc_reviews_score_avg'] = sort_direction;
+      // minimum of 3 reviews needed
+      searchObj['calc_reviews_score_count'] = {
+        $gte: 3,
+      };
+    }
+    else if (sort_by === 'total_reviews') {
+      sortObj['calc_reviews_score_count'] = sort_direction;
+    }
+    else if (sort_by === 'players_beaten') {
```

Comment:
As is, calc_records_count will store the number of times the record has been beaten. If we want the number of players that beat the level then we'll have to use the Stat model instead of the Record model

Reply:
ah ok i need to redo this then.